extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php

```diff
@@ -15,6 +15,159 @@ class SEOTweaksHooksHelper {
+    }
+
+    /**
+     * Return first image from an article, first check for infobox images,
+     * If not found, take the biggest as possible:
+     * first minimal recommended image size from facebook, and if not found, take minimal requirement
+     * @param $title
+     * @return null|Title
+     */
+    static protected function getFirstArticleImage( $title ) {
+        $retTitle = self::getFirstArticleImageLargerThan( $title, self::MIN_WIDTH, self::MIN_HEIGHT, "ImageServingDriverInfoboxImageNS" );
+
+        if ( !empty( $retTitle ) ) {
+            return $retTitle;
+        }
+        $retTitle = self::getFirstArticleImageLargerThan( $title, self::PREF_WIDTH, self::PREF_HEIGHT );
+
+        if ( !empty( $retTitle ) ) {
+            return $retTitle;
+        }
+
+        return self::getFirstArticleImageLargerThan( $title, self::MIN_WIDTH, self::MIN_HEIGHT );
+    }
+
+    /**
+     * Return first image from an article, matched criteria
+     * @param $title
+     * @param $width
+     * @param $height
+     * @param null $driverName
+     * @return null|\Title
+     */
+    static protected function getFirstArticleImageLargerThan( $title, $width, $height, $driverName = null ) {
+        $imageServing = new ImageServing( [ $title->getArticleID() ], $width, $height );
+        $out = $imageServing->getImages( 1, $driverName);
+        return self::createTitleFromResultArray( $out );
+    }
+
+    /**
+     * @desc Creates a Title object from array of file names
+     * @param $out array of file names
+     * @return null|\Title
+     */
+    static protected function createTitleFromResultArray( $out ) {
+        if ( !empty( $out ) ) {
+            ///used reset instead direct call because we can get hashmap from ImageServing driver.
+            $first = reset( $out );
+            $name = $first[ 0 ][ 'name' ];
+            return Title::newFromText( $name, NS_FILE );
+        }
+        return null;
+    }
+
+    /**
+     * @param $meta
+     * @param $title Title
+     * @return bool
+     */
+    static public function onOpenGraphMetaHeaders( &$meta, $title ) {
+        global $wgMemc;
+        if ( !empty( $title ) && $title instanceof Title && !$title->isMainPage() ) {
+            $namespace = $title->getNamespace();
+            if ( $namespace == NS_USER ) {
+                return true;
+            }
+            $cacheKey = self::makeOpenGraphKey( $title );
+            $imageUrl = $wgMemc->get( $cacheKey );
+
+            //if no image in memcache
+            if ( empty( $imageUrl ) ) {
+                if ( $namespace != NS_FILE ) {
+                    $title = self::getFirstArticleImage( $title );
+                }
+                if ( !empty( $title ) ) {
+                    $file = wfFindFile( $title );
+                    if ( !empty( $file ) ) {
+                        $thumb = self::getResizeImageUrlIfLargerThanMax( $file );
+                        if ( !empty( $thumb ) ) $meta[ "og:image" ] = $thumb;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Add curly braces around the nested statement(s). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') 2 statements were found on this line. Reformat the code to have only one statement per line. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS122)